### PR TITLE
#5485 Add sender email validation when sending signed messages

### LINKED ISSUE #5485: Add check for sender alias
ref - https://github.com/FlowCrypt/flowcrypt-security/issues/265

### PR DESCRIPTION
**Pull request by martgil**
This PR adds a sender email validation when sending signed messages.

close #5485

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by martgil** [comment]
Hello @sosnovsky, It's still in a draft state, my bad for not marking it earlier. I'm just updating the tests close enough for you to review after. Thanks!

**Comment by martgil** [comment]
Hi @sosnovsky sorry for the quetions. I tested the test case that fails on the CI manually and it works on a non-testing environment, my manual testing includes using the test private key that in the consts.ts which I updated and it has the user id that I added. However, on the CI test and when running npm run test, I suspect that its using some other private key for flowcrypt.compatibility account... 

I only updated the private key at https://github.com/FlowCrypt/flowcrypt-browser/blob/3358e3584ca58e253adf874862c4dc5a69caa66e/test/source/tests/tooling/consts.ts#L2092...

Do you have any idea where should I updated it? tried finding other files that potentially referenced to it but still no good.

**Comment by sosnovsky** [comment]
Hi @martgil, maybe test account loads key from backup email message - https://github.com/FlowCrypt/flowcrypt-browser/blob/master/test/source/mock/google/exported-messages/message-export-188923a75165a3c8.json?

**Comment by martgil** [comment]
@sosnovsky This one is ready for review. Thanks!

**Comment by martgil** [comment]
> Hi @martgil, maybe test account loads key from backup email message - https://github.com/FlowCrypt/flowcrypt-browser/blob/master/test/source/mock/google/exported-messages/message-export-188923a75165a3c8.json?

Thanks @sosnovsky It actually is, after some time of debugging I'm able to find it when I analyze the `.har` file I downloaded to see the network activity which reveals to me the `threadId` which points to this and to the other file that I must update. Thanks a lot for your help.

**Comment by sosnovsky** [comment]
Hi @martgil, is it ready for review or you want to change something here?

**Comment by martgil** [comment]
I’m still in the moment of fixing the failing test. I think i will let you
later when I fixed it so that you wont be needing to review it more one
more time. Thanks!

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 1]
```diff
@@ -80,4 +89,19 @@ export class GeneralMailFormatter {
     // throws ComposerResetBtnTrigger when user closes pass phrase dialog without entering
     return await view.storageModule.decryptSenderKey(parsedSenderPrv);
   };
+
+  private static isSenderInKeyUserIds = async (view: ComposeView, signingKey?: ParsedKeyInfo | undefined): Promise<boolean> => {
```
Comment (sosnovsky) [inline extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts:93]: Probably we can make `signingKey` non-optional, as there is no reason to check sender in non-existent key.

Also in line 97 there is a check `senderEmail !== view.acctEmail && signingKey?.keyInfo.emails`, and if `signingKey` is `undefined` then we `return true` which is incorrect, as key is undefined and can't include sender

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 2]
```diff
@@ -80,4 +89,19 @@ export class GeneralMailFormatter {
...
+    const senderEmail = view.senderModule.getSender();
+    const storage = await AcctStore.get(view.acctEmail, ['sendAs']);
+    const emailAliases = Object.keys(storage.sendAs || {});
```
Comment (sosnovsky) [inline extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts:96]: these 2 lines can be moved inside `if` statement, as we don't need to fetch aliases when `if` check returns `false`

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 3]
```diff
@@ -80,4 +89,19 @@ export class GeneralMailFormatter {
...
+    if (senderEmail !== view.acctEmail && signingKey?.keyInfo.emails) {
+      const senderAliasIndex = emailAliases.indexOf(view.acctEmail);
+      if (senderAliasIndex !== -1) {
+        emailAliases.splice(senderAliasIndex, 1);
+      }
+      const emailsInSigningKey = signingKey.keyInfo.emails;
+      return emailsInSigningKey.some(email => emailAliases.includes(email));
```
Comment (sosnovsky) [inline extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts:103]: hmm, not sure if it's correct:

for example, we have user with `view.acctEmail = test.user@gmail.com`, it has several aliases - `test.user@gmail.com`, `testuser@gmail.com`, `test+user@gmail.com`.

and signing key has `signingKey.keyInfo.emails = ['test.user@gmail.com', 'testuser@gmail.com']`

then user tries to send message using alias `test+user@gmail.com` and there won't be any error as `emailsInSigningKey` will include `testuser@gmail.com`

or I'm missing something?

Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts:103]: Ah yes, thanks for noticing. I thought those filters in email only applies in inboxes... There wouldn't be any error indeed in that case you have mentioned. I'll update it too.

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts (hunk 1 above)
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: Thanks, I made the necessary updates here.

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts (hunk 2 above)
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: Right, thanks!

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts (hunk 3 above)
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: updated. thanks!

**Comment by martgil** [comment]
@sosnovsky ready for review. thanks!

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 4]
```diff
@@ -80,4 +89,24 @@ export class GeneralMailFormatter {
...
+      if (senderAliasIndex !== -1) {
+        emailAliases.splice(senderAliasIndex, 1);
+      }
+      const emailsInSigningKey = signingKey.keyInfo.emails || [];
+      let filteredEmailsInSigningKey: string[] = [];
+      for (const email of emailsInSigningKey) {
+        filteredEmailsInSigningKey.push(email.replace(/(.+)(?=\+).*(?=@)/, '$1'));
+      }
+      filteredEmailsInSigningKey = filteredEmailsInSigningKey.filter((value, index) => filteredEmailsInSigningKey.indexOf(value) === index);
+      return emailAliases.every(email => filteredEmailsInSigningKey.includes(email));
```
Comment (sosnovsky) [inline extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts:108]: As I understand, currently you fetch all user aliases from `AcctStore`, then get all emails from signing key and check if signing key emails contain all users aliases?

But I think we don't need to check for all user aliases, as user uses only 1 alias for sending message, so correct check should be just:
- get all emails from signing key
- check if these emails include `senderEmail`

Probably it should be enough.

P.S. Can you please explain what filtering you perform here (as it's difficult to understand without additional comments, feel free to add them to the code):
```ts
let filteredEmailsInSigningKey: string[] = [];
for (const email of emailsInSigningKey) {
    filteredEmailsInSigningKey.push(email.replace(/(.+)(?=\+).*(?=@)/, '$1'));
}
filteredEmailsInSigningKey = filteredEmailsInSigningKey.filter((value, index) => filteredEmailsInSigningKey.indexOf(value) === index);
```

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 5]
```diff
@@ -80,4 +89,24 @@ export class GeneralMailFormatter {
     // throws ComposerResetBtnTrigger when user closes pass phrase dialog without entering
     return await view.storageModule.decryptSenderKey(parsedSenderPrv);
   };
+
+  private static isSenderInKeyUserIds = async (view: ComposeView, signingKey: ParsedKeyInfo): Promise<boolean> => {
+    const senderEmail = view.senderModule.getSender();
+    if (senderEmail !== view.acctEmail && signingKey.keyInfo.emails) {
```
[the remark this answers is not in the document]
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: I've corrected this one. thanks!

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts (hunk 4 above)
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: Yes, Roma, my bad. I have updated it and re-considered the aliases with filters "+" and sanitize them for better matching.

**Comment by martgil** [comment]
Hello @sosnovsky I updated the requested changes and reviewed it. It looks good to me -- can you give it a second look?

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 6]
```diff
@@ -80,4 +89,19 @@ export class GeneralMailFormatter {
...
+    const senderEmail = view.senderModule.getSender();
+    const emailsInSigningKey = signingKey.emails;
+    if (!emailsInSigningKey) {
+      return false;
+    } else if (senderEmail !== view.acctEmail) {
+      // Transpose the email to its base form when a filter is present. This is important
+      // to consider when an email with a filter is present on the signingKey, as
+      // technically it is the same email despite the filter.
+      const baseSenderEmail = senderEmail.indexOf('+') !== -1 ? senderEmail.replace(/(.+)(?=\+).*(?=@)/, '$1') : senderEmail;
+      return emailsInSigningKey.some(email => email.includes(baseSenderEmail));
```
[the remark this answers is not in the document]
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: Thanks Roma, I've made a quick search and found out it was formally called 'email tags' - https://support.google.com/manufacturers/answer/6184604. I'll be updating the comments. thanks!

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 7]
```diff
@@ -42,6 +42,11 @@ export class GeneralMailFormatter {
       const signingKey = await GeneralMailFormatter.chooseSigningKeyAndDecryptIt(view, senderKis);
       if (!signingKey) {
         throw new UnreportableError("Could not find account's key usable for signing this plain text message");
+        /* eslint-disable @typescript-eslint/no-non-null-assertion */
+      } else if (!(await GeneralMailFormatter.isSenderInKeyUserIds(view, signingKey.key))) {
+        throw new UnreportableError(
+          `Could not sign this encrypted message. The sender email ${view.senderModule.getSender()} isn't present in the signing key's user ids`
+        );
       }
       /* eslint-disable @typescript-eslint/no-non-null-assertion */
```
[the remark this answers is not in the document]
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: understood. thanks!

File: extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts [hunk 8]
```diff
@@ -42,6 +42,11 @@ export class GeneralMailFormatter {
       const signingKey = await GeneralMailFormatter.chooseSigningKeyAndDecryptIt(view, senderKis);
       if (!signingKey) {
         throw new UnreportableError("Could not find account's key usable for signing this plain text message");
+        /* eslint-disable @typescript-eslint/no-non-null-assertion */
+      } else if (!(await GeneralMailFormatter.isSenderInKeyUserIds(view, signingKey.key))) {
```
[the remark this answers is not in the document]
Comment (martgil) [reply extension/chrome/elements/compose-modules/formatters/general-mail-formatter.ts]: will remove it, thanks!

**Comment by martgil** [comment]
Hi @sosnovsky, I have completed all the necessary changes for this PR, and it's all good now.

**Review by sosnovsky — APPROVED** [review]
All good now, thanks!

**Pull request merged**